internal/pkg/descriptor/parse.go

```diff
@@ -0,0 +1,314 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package descriptor
+
+import (
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/operator-framework/operator-sdk/internal/annotations"
+	"github.com/operator-framework/operator-sdk/internal/util/k8sutil"
+
+	olmapiv1alpha1 "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/pkg/errors"
+	"k8s.io/gengo/types"
+)
+
+const csvgenPrefix = annotations.SDKPrefix + ":gen-csv:"
+
+type descriptorType = string
+
+const (
+	typeSpec   descriptorType = "spec"
+	typeStatus descriptorType = "status"
+)
+
+type descriptor struct {
+	include     bool
+	parentType  *types.Type
+	member      types.Member
+	descType    descriptorType
+	description string
+	displayName string
+	path        string
+	xdesc       []string
+}
+
+func sortDescriptors(ds []descriptor) []descriptor {
+	sort.Slice(ds, func(i, j int) bool {
+		return ds[i].path < ds[j].path
+	})
+	return ds
+}
+
+type parsedCRDDescriptions struct {
+	descriptors []descriptor
+	displayName string
+	resources   []olmapiv1alpha1.APIResourceReference
+}
+
+func sortResources(rs []olmapiv1alpha1.APIResourceReference) []olmapiv1alpha1.APIResourceReference {
+	sort.Slice(rs, func(i, j int) bool {
+		return rs[i].Kind < rs[j].Kind
+	})
+	return rs
+}
+
+func wrapParseErr(err error) error {
+	return errors.Wrap(err, "error parsing csv-gen annotation")
+}
+
+// parseCSVGenAnnotations parses all descriptor annotations from comments,
+// each of which should contain one spec.customresourcedefinitions.owned entry.
+// field Once all comments have been parsed, the entry is added to a
+// parsedCRDDescriptions.
+func parseCSVGenAnnotations(comments []string) (pd parsedCRDDescriptions, err error) {
```

Comment:
@estroz Can we add unit tests for `parseCSVGenAnnotations()`?
That and `mergeChildDescriptorPaths()` which I'm still trying to understand.

Ideally we'd have tests for `GetCRDDescriptorForGVK()` too but I imagine that is probably more suited to being an e2e test among other tests for `operator-sdk olm-catalog gen-csv` command.

Reply:
`GetCRDDescriptorForGVK()` is tested (somewhat) by CSV scaffold unit tests. I will be submitting PR's for `internal/pkg/scaffold/olm-catalog` unit tests soon, but because the functionality added here is complex there should be unit tests in this PR. I'll add them before merging.